### PR TITLE
Add bucket extraction function

### LINKED ISSUE #2973: Make a 'trunc' extraction function for numeric dimensions
I would like to be able to express the following extraction function:

```javascript
{
  "extractionFn": {
    "function": "function(d){d=Number(d); if(isNaN(d)) return 'null'; return Math.floor((d - 0.5) / 2) * 2 + 0.5;}",
    "type": "javascript"
  },
  "dimension": "height_bucket",
  "outputName": "HeightBucket",
  "type": "extraction"
}
```

Without using JavaScript (make it native)

I suggest an extraction function that look like this: 

```javascript
{
  "extractionFn": {    
    "type": "trunc", // or "floor"
    "size": 2, // optional, defaults to 1
    "offset": 0.5 // optional defaults to 0
  },
  "dimension": "height_bucket",
  "outputName": "HeightBucket",
  "type": "extraction"
}
```

Which would translate to: `function(d){d=Number(d); if(isNaN(d)) return 'null'; return Math.floor((d - offset) / size) * size + offset;}`

This is needed to implement queries like:

```sql
SELECT TRUNC(commentLength, 10) AS "commentLengthBucketsOf10"
FROM "wikipedia"
GROUP BY TRUNC(commentLength, 10)
```

### PR DESCRIPTION
**Pull request by acslk**
Fixes #2973 
Adds bucket extraction functions that bucket numeric dimension values. It converts all value in the range of  [x * size + offset, (x+1) * size + offset) to the value (x * size + offset) for any integer x. The default values for size and offset, if not specified in the json, are 1 and 0.

### CODE REVIEW

File: processing/src/main/java/io/druid/segment/IndexMerger.java [hunk 1]
```diff
@@ -341,31 +341,21 @@ public File merge(
...
-        ),
-        new Function<String, String>()
-        {
-          @Override
-          public String apply(@Nullable String input)
-          {
-            return input;
-          }
-        }
+    final List<String> mergedMetrics = mergeIndexed(
```
Comment (drcrallen) [inline processing/src/main/java/io/druid/segment/IndexMerger.java]: Why does this need to change?

Comment (acslk) [inline processing/src/main/java/io/druid/segment/IndexMerger.java]: It's just an unrelated refactor, it seems like this transform function does not do anything after a previous change, so the transform is not nessasary.

File: processing/src/main/java/io/druid/query/extraction/BucketExtractionFn.java [hunk 2]
```diff
@@ -0,0 +1,109 @@
+package io.druid.query.extraction;
```
Comment (fjy) [inline processing/src/main/java/io/druid/query/extraction/BucketExtractionFn.java]: missing header

File: processing/src/test/java/io/druid/query/extraction/BucketExtractionFnTest.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
+package io.druid.query.extraction;
```
Comment (fjy) [inline processing/src/test/java/io/druid/query/extraction/BucketExtractionFnTest.java]: header

**Comment by fjy** [comment]
@acslk this is missing user documentation

File: processing/src/main/java/io/druid/segment/IndexMerger.java (hunk 1 above)
Comment (gianm) [inline processing/src/main/java/io/druid/segment/IndexMerger.java]: Please do this in another PR, it makes life easier if a commit contains a set of related changes.

**Comment by gianm** [comment]
@fjy do you mean in dimensionspecs.md?

File: processing/src/main/java/io/druid/query/extraction/BucketExtractionFn.java (hunk 2 above)
Comment (acslk) [inline processing/src/main/java/io/druid/query/extraction/BucketExtractionFn.java]: does header refer to the "Licensed to Metamarkets Group..." comment?

**Comment by fjy** [comment]
@gianm yeah

File: processing/src/main/java/io/druid/query/extraction/BucketExtractionFn.java (hunk 2 above)
Comment (fjy) [inline processing/src/main/java/io/druid/query/extraction/BucketExtractionFn.java]: yes

File: docs/content/querying/dimensionspecs.md [hunk 4]
```diff
@@ -414,6 +414,23 @@ or without setting "locale" (in this case, the current value of the default loca
 }
 ```
 
+### Bucket Extraction Function
+
+Bucket extraction funciton is used to bucket numerical values in each range of the given size by converting them to the same base value. Non numeric values are converted to empty strings.
```
Comment (vogievetsky) [inline docs/content/querying/dimensionspecs.md]: > Non numeric values are converted to empty strings.

I would rather this said "Non numeric values are converted to null." this is the same if I understand since `"" == null` in Druid but to someone coming in "null" make a lot more sense than "empty string" in this context.

File: processing/src/test/java/io/druid/query/extraction/BucketExtractionFnTest.java [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+  @Test
+  public void testApply()
+  {
+    BucketExtractionFn extractionFn1 = new BucketExtractionFn(100.0, 0.5);
+    Assert.assertEquals("1200.5", extractionFn1.apply("1234.99"));
+    Assert.assertEquals("0.5", extractionFn1.apply("1"));
+    Assert.assertEquals("0.5", extractionFn1.apply("100"));
+    Assert.assertEquals("500.5", extractionFn1.apply(501));
+    Assert.assertEquals("-399.5", extractionFn1.apply("-325"));
```
Comment (vogievetsky) [inline processing/src/test/java/io/druid/query/extraction/BucketExtractionFnTest.java]: Would be great to see a test for input that is non numeric, specifically `""` and `"blah blah"` also `"1e2"` would be an interesting test cases

**Comment by gianm** [comment]
👍  LGTM
